compiler/src/main/java/com/squareup/anvil/compiler/ClassScanner.kt

```diff
@@ -1,98 +1,177 @@
+      contributesMultibindingFqName -> HINT_MULTIBINDING_PACKAGE_PREFIX
+      contributesSubcomponentFqName -> HINT_SUBCOMPONENTS_PACKAGE_PREFIX
+      else -> throw AnvilCompilationException(message = "Cannot find hints for $annotation.")
+    }
+
+    val packageDescriptor = module.getPackage(FqName(packageName))
+    return generateSequence(listOf(packageDescriptor)) { subPackages ->
+      subPackages
+        .flatMap { it.subPackages() }
+        .ifEmpty { null }
+    }
+      .flatMap { it.asSequence() }
+      .flatMap {
+        it.memberScope.getContributedDescriptors(DescriptorKindFilter.VALUES)
       }
+      .filterIsInstance<PropertyDescriptor>()
+      .mapNotNull { GeneratedProperty.fromDescriptor(it) }
+      .groupBy { property -> property.baseName }
+      .values
+      .asSequence()
+      .map { ContributedHint(it, annotation) }
   }
 }
 
 private fun PackageViewDescriptor.subPackages(): List<PackageViewDescriptor> = memberScope
   .getContributedDescriptors(DescriptorKindFilter.PACKAGES)
   .filterIsInstance<PackageViewDescriptor>()
 
-private class ContributedHint(properties: List<PropertyDescriptor>) {
-  val reference by lazy(NONE) {
-    properties
-      .bySuffix(REFERENCE_SUFFIX)
-      .toClassDescriptor()
-  }
+sealed class GeneratedProperty(
+  val descriptor: PropertyDescriptor,
+  val baseName: String
+) {
+  class ReferenceProperty(
+    descriptor: PropertyDescriptor,
+    baseName: String
+  ) : GeneratedProperty(descriptor, baseName)
 
-  val scope by lazy(NONE) {
-    properties
-      .bySuffix(SCOPE_SUFFIX)
-      .toClassDescriptor()
-  }
+  class ScopeProperty(
+    descriptor: PropertyDescriptor,
+    baseName: String
+  ) : GeneratedProperty(descriptor, baseName)
 
-  private fun List<PropertyDescriptor>.bySuffix(suffix: String): PropertyDescriptor = first {
-    it.name.asString()
-      .endsWith(suffix)
+  companion object {
+    fun fromDescriptor(descriptor: PropertyDescriptor): GeneratedProperty? {
+      // For each contributed hint there are several properties, e.g. the reference itself
+      // and the scopes. Group them by their common name without the suffix.
+      val name = descriptor.name.asString()
+
+      return when {
+        name.endsWith(REFERENCE_SUFFIX) ->
+          ReferenceProperty(descriptor, name.substringBeforeLast(REFERENCE_SUFFIX))
+        name.contains(SCOPE_SUFFIX) -> {
+          // The old scope hint didn't have a number. Now that there can be multiple scopes
+          // we append a number for all scopes, but we still need to support the old format.
+          val indexString = name.substringAfterLast(SCOPE_SUFFIX)
+          if (indexString.toIntOrNull() != null || indexString.isEmpty()) {
+            ScopeProperty(descriptor, name.substringBeforeLast(SCOPE_SUFFIX))
+          } else {
+            null
+          }
+        }
+        else -> null
+      }
+    }
   }
+}
+
+// TODO Repeatable: Cleanup this class with what's actually required.
```

Comment:
Not sure if you use github to track your issues, but worthwhile to note your ticket number here if you do.

Reply:
I have to clean up all deprecated methods and TODOs later. This PR would otherwise explode even further.